File: python/relicpy/shapes/claude.py
"""Claude Code transcripts — ~/.claude/projects/<encoded>/<uuid>.jsonl and its
subagent / workflow-agent children.

One JSON object per line, and the WHOLE file is transcript — unlike Codex, where only
one record type is.
"""

from __future__ import annotations

import json
import os
import re

from ..models import ParsedEvent, ParsedFile
from ..types import as_obj, block_role, flatten_content, s, tree_key_of, truncate, uid_of

# Roles whose text is worth full-text indexing. UI/state events are counted, not indexed.
INDEXED = {"user", "assistant", "system"}

_UUID36 = re.compile(r"^[0-9a-f-]{36}$")


def parse(file_path: str) -> ParsedFile:
    file_key = os.path.basename(file_path)
    uid_key = tree_key_of(file_path)
    events: list[ParsedEvent] = []
    lines = bad_lines = seq = 0
    cwd = model = None
    session_uuid = re.sub(r"\.jsonl$", "", file_key)
    started_at = ended_at = description = title = git_branch = None

    with open(file_path, "r", encoding="utf-8", errors="replace") as fh:
        for line in fh:
            trimmed = line.strip()
            if not trimmed:
                continue          # seq counts NON-EMPTY lines only — see `show`
            lines += 1
            seq += 1

            try:
                rec = as_obj(json.loads(trimmed))
            except Exception:
                bad_lines += 1
                continue
            if rec is None:
                bad_lines += 1
                continue

            typ = s(rec.get("type")) or "unknown"

            # cwd comes from the session's own field, never decoded from the directory
            # name: the encoding maps BOTH "/" and "." to "-", so it is lossy.
            # `cwd` (first seen) identifies the SESSION and picks its shard.
            # `line_cwd` is what THIS record said, and rides on the event — a session
            # that changes repos mid-flight had 201 events unreachable by cwd before.
            line_cwd = s(rec.get("cwd"))
            if not cwd:
                cwd = line_cwd
            if not _UUID36.match(session_uuid):
                session_uuid = s(rec.get("sessionId")) or session_uuid

            ts = s(rec.get("timestamp"))
            if ts:
                if not started_at:
                    started_at = ts
                ended_at = ts

            if not git_branch:
                git_branch = s(rec.get("gitBranch"))
            if typ == "ai-title":
                t = s(rec.get("aiTitle"))
                if t:
                    # Overwrite, never keep-first: the record repeats on nearly every
                    # turn and the title is refined as the session goes on.
                    title = t

            if typ not in INDEXED:
                continue

            msg = as_obj(rec.get("message"))
            if not model and msg:
                model = s(msg.get("model"))
            raw = (msg.get("content") if msg else None)
            if raw is None:
                raw = rec.get("content", "")
            text = flatten_content(raw).strip()
            if not text:
                continue

            # Label by what the block ACTUALLY is, not by the envelope that carried it.
            role = block_role(raw) or (s(msg.get("role")) if msg else None) or typ

            if not description and role == "user":
                description = truncate(text, 200)
            events.append(ParsedEvent(
                uid=uid_of("claude", uid_key, seq),    # root-independent by design
                seq=seq, role=role, ts=ts, text=truncate(text), cwd=line_cwd,
            ))

    return ParsedFile(
        session_uuid=session_uuid, cwd=cwd, model=model, events=events,
        lines=lines, bad_lines=bad_lines, started_at=started_at, ended_at=ended_at,
        description=description, title=title, git_branch=git_branch,
    )

File: src/shapes/claude.ts
import { createHash } from "node:crypto";
import { createReadStream } from "node:fs";
import { createInterface } from "node:readline";
import { basename } from "node:path";
import { asObj, str, truncate, flattenContent, blockRole, uidOf, type ParsedEvent, type ParsedFile, type Parser } from "../types.js";

/** Roles whose text is worth full-text indexing. UI/state events are counted, not indexed. */
const INDEXED = new Set(["user", "assistant", "system"]);

/**
 * Claude Code transcripts — ~/.claude/projects/<encoded>/<uuid>.jsonl and its
 * subagent / workflow-agent children. One JSON object per line, and the whole file
 * is transcript (unlike Codex, where only one record type is).
 */
export const parseClaude: Parser = async (filePath) => {
  const fileKey = basename(filePath);
  const events: ParsedEvent[] = [];
  const typeCounts: Record<string, number> = {};
  let lines = 0, badLines = 0, seq = 0;
  let cwd: string | null = null, model: string | null = null;
  let sessionUuid = fileKey.replace(/\.jsonl$/, "");
  let startedAt: string | null = null, endedAt: string | null = null;
  let description: string | null = null;

  const rl = createInterface({ input: createReadStream(filePath, "utf8") });
  for await (const line of rl) {
    const trimmed = line.trim();
    if (!trimmed) continue;          // seq counts NON-EMPTY lines only — see `show`
    lines++; seq++;

    let parsed: unknown;
    try { parsed = JSON.parse(trimmed); } catch { badLines++; continue; }
    const rec = asObj(parsed);
    if (!rec) { badLines++; continue; }

    const type = str(rec.type) ?? "unknown";
    typeCounts[type] = (typeCounts[type] ?? 0) + 1;

    // cwd comes from the session's own field, never decoded from the directory name:
    // the encoding maps BOTH "/" and "." to "-", so it is lossy and not reversible.
    if (!cwd) cwd = str(rec.cwd);
    if (!/^[0-9a-f-]{36}$/.test(sessionUuid)) sessionUuid = str(rec.sessionId) ?? sessionUuid;

    const ts = str(rec.timestamp);
    if (ts) { if (!startedAt) startedAt = ts; endedAt = ts; }

    if (!INDEXED.has(type)) continue;

    const msg = asObj(rec.message);
    if (!model) model = str(msg?.model);
    const raw = msg?.content ?? rec.content ?? "";
    const text = flattenContent(raw).trim();
    if (!text) continue;

    // Label by what the block ACTUALLY is, not by the envelope that carried it.
    //
    // Claude delivers a tool result as a `user` message whose content is a
    // tool_result block — so taking the envelope role at face value files machine
    // output as something the human said. Measured on one shard before this fix:
    // 68.5% of indexed text was tool-output-shaped while only 3.7% carried the
    // tool_result role. That buries human prose under command output in every
    // ranked search, and it is why a search for a term printed by a tool ranks the
    // tool's own output first.
    const role = blockRole(raw) ?? str(msg?.role) ?? type;

    if (!description && role === "user") description = truncate(text, 200);
    events.push({
      uid: uidOf("claude", fileKey, seq),   // path-independent by design
      seq, role, ts, text: truncate(text),
    });
  }

  return { sessionUuid, cwd, model, events, lines, badLines, typeCounts, startedAt, endedAt, description };
};
